core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -72,20 +96,64 @@ public FilteredQueryBuilder queryName(String queryName) {
         return this;
     }
 
+    /** Returns the name of this query. */
+    public String queryName() {
+        return queryName;
+    }
+
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject(NAME);
-        if (queryBuilder != null) {
-            builder.field("query");
-            queryBuilder.toXContent(builder, params);
+    protected boolean doEquals(FilteredQueryBuilder other) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder) &&
+                Objects.equals(boost, other.boost) &&
+                Objects.equals(queryName, other.queryName);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(queryBuilder, filterBuilder, boost, queryName);
+    }
+
+    @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+
+        Query query = (queryBuilder == null) ? null : queryBuilder.toQuery(parseContext);
+
+        // parsed internally, but returned null during parsing...
+        if (query == null) {
+            return null;
         }
-        if (filterBuilder != null) {
-            builder.field("filter");
-            filterBuilder.toXContent(builder, params);
+
+        Query filter = (filterBuilder == null) ? null : filterBuilder.toQuery(parseContext);
+        Query result = null;
+
+        if (filter == null) {
+            result = query;
+        } else if (Queries.isConstantMatchAllQuery(filter)) {
+            // this is an instance of match all filter, just execute the query
+            result = query;
```

Comment:
we lose the boost here and in the case before? maybe the setBoost should be at the end of the if?